scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -147,270 +155,92 @@ object ExecutionOptimizationRules {
-                      .optimize
-                      .map(_.map(ZipMap.MapRight(m.mapFn)))
-                  }
-
-                  go(m).map(fix(_))
-              }
-          }
+      def map[A, B](ex: FlattenedZip[A], fn: A => B): FlattenedZip[B] = ex match {
+        case NonWrite(nonWrite) => NonWrite(nonWrite.map(fn))
+        case Write(write) => Write(mapWrite(write, fn))
+        case Composed(write, nonWrite, compose) => Composed(write, nonWrite, ComposedMapFn(compose, fn))
       }
 
-      object Many {
-        /**
-         * This is a handy constructor function that builds the SubTypes instance for us
-         */
-        def apply[Ex[x] <: Execution[x], A, B](head: Ex[A], rest: FlattenedZip[Execution, B]): Many[Ex, A, B, (A, B)] =
-          Many(head, rest, SubTypes.fromSubType[(A, B), (A, B)])
+      def zip[A, B](left: FlattenedZip[A], right: FlattenedZip[B]): FlattenedZip[(A, B)] = (left, right) match {
+        case (left@NonWrite(_), right@NonWrite(_)) =>
+          NonWrite(left.nonWrite.zip(right.nonWrite))
+        case (left@NonWrite(_), right@Write(_)) =>
+          Composed(right.write, left.nonWrite, Swap[B, A]())
+        case (left@NonWrite(_), right@Composed(_, _, _)) =>
+          zipNonWriteComposed(left, right)
+
+        case (left@Write(_), right@NonWrite(_)) =>
+          Composed(left.write, right.nonWrite, Identity[(A, B)]())
+        case (left@Write(_), right@Write(_)) =>
+          Write(mergeWrite(left.write, right.write))
+        case (left@Write(_), right@Composed(_, _, _)) =>
+          zipWriteComposed(left, right)
+
+        case (left@Composed(_, _, _), right@NonWrite(_)) =>
+          map(zipNonWriteComposed(right, left), Swap[B, A]())
+        case (left@Composed(_, _, _), right@Write(_)) =>
+          map(zipWriteComposed(right, left), Swap[B, A]())
+        case (left@Composed(_, _, _), right@Composed(_, _, _)) =>
+          Composed(mergeWrite(left.write, right.write), left.nonWrite.zip(right.nonWrite),
+            ComposedFn(left.compose, right.compose))
       }
 
-      /**
-       * This represents pushing all the map functions after the zips, this is needed to
-       * handle the internal mapping functions we do to reverse the order of zips and get
-       * the WriteExecution moves all the way over to the head
-       */
-      final case class MapZip[Ex[x] <: Execution[x], A, B](flat: FlattenedZip[Ex, A], mapFn: A => B) extends FlattenedZip[Ex, B] {
-        type ExH = flat.ExH
-
-        def executions: ExH = flat.executions
-        def toExecution(ex: ExH): Execution[B] =
-          Mapped(flat.toExecution(ex), mapFn)
-
-        def zip[C](that: FlattenedZip[Execution, C]): FlattenedZip[Execution, (B, C)] =
-          MapZip(flat.zip(that), ZipMap.MapLeft[A, C, B](mapFn))
+      private def zipNonWriteComposed[A, B1, B2, B](left: NonWrite[A], right: Composed[B1, B2, B]): Composed[B1, (B2, A), (A, B)] =
+        Composed(right.write, right.nonWrite.zip(left.nonWrite),
```

Comment:
I wonder if the function get simpler if we do `left.nonWrite.zip(right.nonWrite)` since you have to move the `A` fewer steps to the left?

Reply:
Then you'll have `Composed[B1, (A, B2), (A, B)]` and i.e. `(B1, (A, B2))` => `(A, B)` which is unclear how to untwist simpler.